SchedulingType.FOUR_STAGE: GEMM Full Software Pipelining with Initiation Interval 1, Via Multibuffering

SourishW opened the pull request:
Tested this new pipelined schedule on various tests in wave_gemm_test.py. Also modified wave_gemm_test.py correspondingly. 

Here is a representation of the type of schedule generated from a (GR, LW, LR, MMA) GEMM kernel:
```

a_0 = READ_GLOBAL 0

WRITE_SHARED a_0 SM[0]
a_1 = READ_GLOBAL 1
lds_barrier

b_0 = READ_SHARED SM[0]
WRITE_SHARED a_1 SM[1]
a_2 = READ_GLOBAL 2
lds_barrier

for i = 0 to N -3:
    COMPUTE b_i
    b_{i+1} = READ_SHARED SM[i+1 %2]
    // Work in progress is removing the immediate subsequent lds_barrier
    (lds_barrier)
    WRITE_SHARED a_{i+2} SM[i%2]
    a_{i+3} = READ_GLOBAL k+3
    lds_barrier

COMPUTE b_{n-2}
b_{n-1} = READ_SHARED SM[n-1 %2]
WRITE_SHARED a_{n} SM[n % 2]

COMPUTE b_{n-1}
b_{n} = READ_SHARED SM[n %2]

COMPUTE b_n
```

Thank you in advance for reviewing

raikonenfnu commented on wave_lang/kernel/wave/scheduling/gemm_four_stage_pipelined_scheduling.py:
```diff
@@ -0,0 +1,144 @@
+    LOCAL_STORE = 1
+    LOCAL_LOAD = 2
+    COMPUTE = 3
+    DOESNT_MATTER = -1
```
Any better name aside from DOESNT_MATTER? np if not

SourishW replied:
Oh I forgot about that, I was using a temporary name until I thought of a better one, I will fix it

SourishW replied:
I fixed this btw forgot to reply

raikonenfnu commented on wave_lang/kernel/wave/scheduling/gemm_four_stage_pipelined_scheduling.py:
```diff
@@ -0,0 +1,144 @@
+        """
+        Returns the number of stages in the kernel of the pipelined loop.
+        """
+        return 4
```
Why not use the mathematical formula? The math formula is important because it will only generate `4`, if we are able to do scheduling properly

SourishW replied:
Sure, will fix

raikonenfnu commented on wave_lang/kernel/wave/scheduling/multi_buffering.py:
```diff
@@ -24,9 +24,36 @@
+    new_shape = []
+    new_distributed_shape = []
+
+    for i, dim in enumerate(original_buffer.shape):
+        if i in reduction_dim_indices:
+            # Keep reduction dimensions as is
+            new_shape.append(dim)
+            new_distributed_shape.append(original_buffer.distributed_shape[i])
+        else:
+            # Increase non-reduction dimensions
+            new_shape.append(dim * buffer_count)
+            new_distributed_shape.append(
+                original_buffer.distributed_shape[i] * buffer_count
+            )
+
+    original_buffer.update_arg(0, tuple(new_shape))
+    original_buffer.update_arg(1, tuple(new_distributed_shape))
```
I think this is fine for this PR, but in future PR, maybe would be good to have the expanded dimension the slowest/first dimension instead of the fastest/last dimension. CC: @harsh-nod

harsh-nod replied:
Yes, I would agree that it should be the slowest dimension.

raikonenfnu commented on wave_lang/kernel/wave/scheduling/gemm_four_stage_pipelined_scheduling.py:
```diff
@@ -0,0 +1,144 @@
+        edges: list[Edge],
+        resources: list[int],
+    ) -> None:
+        # assert False
```
dead code here?

SourishW commented on wave_lang/kernel/wave/scheduling/loop_reconstruction.py:
```diff
@@ -42,7 +48,28 @@ class PipelineStage(Enum):
     EPILOGUE = 2
 
 
+def _heuristically_multi_buffer_shared_memory(
```
The modulo case handles this multi_buffering of shared memory after all operations have been already scheduled. IE it applies this almost as a pass afterwards, to find all shared memory and multibuffer it. 

I originally tried to do a similar approach and put it next to MODULO_MULTI_BUFFERED code after the fact, but was having issues finding which loop iterations needed which induction variable, since they're all offset, etc. Including it in loop reconstruction enables me to more conveniently have access to induction_variable that is different for each operation. (ie the induction_variable in loop_reconstruction.py doesnt mean the same thing as induction_variable in multi_buffering.py)

Here since we know in advance which iteration we are on, I decided to update it while generating the operations themselves. It would have been a little trickier to do it the other way. 

Actually btw we can even move modulo code into loop_reconstruction.py later as well.

raikonenfnu replied:
> which loop iterations needed which induction variable

Couldn't we just use `from ..utils.general_utils import get_induction_variable; induction_variable = get_induction_variable(reduction, constraints)`?

SourishW replied:
Those current_induction_variables are a little bit more complicated than just the one you get from reduction and constraints I think. Its different for prologue kernel epilogue. 

If the function add_nodes_by_schedule is called in prologue code section, then current_induction_variables become the prologue 0, 1, 2, etc etc. If called in kernel, then current_induction_variable becomes ARGK+0, 1 etc. If called in epilogue its K-3, K-2, etc

Actually if you trace up code in kernel it actually gets the original induction variable from the exact thing you mention. So we do use that.

harsh-nod replied:
So just trying to understand why is this different from the code that already exists in multi_buffering.py because there the induction variables can be remapped without doing this inside loop reconstruction?

SourishW replied:
I forgot to mention about prologue and epilogue things, thats actually a bigger reason really:

The problem is you can't easily find the prologue and epilogue sections to multibuffer it in multi_buffering.py without adding some really weird attributes or some graph analysis to help identification of these segments later. 

Actually its hard so the current MODULO_MULTI_BUFFERED code doesn't really bother to look at prologue or epilogue either. 

It just finds an iterate block and multibuffers it, not caring about prologue and epilogue pipelining

<img width="645" height="208" alt="image" src="https://github.com/user-attachments/assets/e788b915-9c00-4bfc-9df7-3fced637f3eb" />

You can see evidence of this in the IR too:

<img width="645" height="208" alt="image" src="https://github.com/user-attachments/assets/5ab509e6-510c-454a-9baf-81ddaab1f3b4" />

This whole approach of doing some pipelining then trying tracing your graph to reverse engineer your own pipeline to find prologue kernel and epilogue didn't make sense to me, so I thought why not do your multibuffering while you know what you're producing? Let me know if you have any comments on this @harsh-nod 

Will try to move code for MODULO_MULTI_BUFFERING too soon if thats ok.

raikonenfnu commented:
you have git conflicts btw, @SourishW

harsh-nod commented on lit_tests/kernel/wave/gemm.py:
```diff
@@ -1273,6 +1273,101 @@ def repeat(acc: tkl.Register[M, N, tkl.f32]) -> tkl.Register[M, N, tkl.f32]:
+    # CHECK: vector.store %{{.*}}, %[[VIEW0]][%[[STORE_IDX]], %{{.*}}] : memref<128x36xf16
+
+    # Verify prologue loads from shared memory
+    # CHECK: %[[LOAD_IDX1:.*]] = affine.apply #map4()[%thread_id_x, %thread_id_y]
```
I don't think you should match to the map number exactly. The ordering of the maps and hence the map numbering is arbitrary so it could change.

SourishW replied:
I'm sorry for making this silly mistake, I forgot to replace this matching to affine map with matching to anything when I was replacing with regex. Will try to be more careful with code next time

SourishW replied:
Fixed this and matched to anything. 
` # CHECK: vector.store %{{.*}}, %[[VIEW0]][%[[STORE_IDX]], %{{.*}}] : memref<128x36xf16

    # Verify prologue loads from shared memory
    # CHECK: %[[LOAD_IDX1:.*]] = affine.apply #[[MAP_LOAD1:.*]]()[%thread_id_x, %thread_id_y]`

harsh-nod commented on wave_lang/kernel/wave/scheduling/gemm_four_stage_pipelined_scheduling.py:
```diff
@@ -0,0 +1,148 @@
+            COMPUTE b_i
+            b_{i+1} = READ_SHARED SM[i+1 %2]
+            WRITE_SHARED a_{i+2} SM[i%2]
+            a_{i+3} = READ_GLOBAL k+3
```
Should this be READ_GLOBAL i  + 3 ? With this schedule, there should be no barriers inside the loop right?

SourishW replied:
You still need a barrier in the loop, just not two barriers and not in between the READ_SHARED and WRITE_SHARED segment. 

With double buffering Loop iteration i's write to SM[i%2] has to complete before loop iteration i+1's read from SM[(i+1)+1 %2] ie SM [i%2].

harsh-nod commented on wave_lang/kernel/wave/scheduling/gemm_four_stage_pipelined_scheduling.py:
```diff
@@ -0,0 +1,148 @@
+}
+
+
+def get_scheduling_stage(op: fx.Node) -> GemmFourStageStage:
```
Should the return type be Operation? Also seems like you could reuse this function from prefetch_scheduling.py?

martin-luecke commented on wave_lang/kernel/wave/scheduling/scheduler_utils.py:
```diff
@@ -0,0 +1,47 @@
+from ...ops.wave_ops import get_custom
+import math
+
+ScheduleStage = Enum  # alias
+
+
+def get_scheduling_stage(
+    op: fx.Node, operation_stage_table: dict[ScheduleStage:int]
+) -> ScheduleStage:
+    op_ty = get_custom_operation_type(get_custom(op))
+    assert op_ty is not None, f"get_custom_operation_type returned None for {op}"
+    if op_ty not in operation_stage_table:
+        raise NotImplementedError(f"Cannot find {op_ty} in operation_stage_table")
+    return operation_stage_table[op_ty]
```
The type annotation of `operation_stage_table` is wrong. In [gemm_four_stage_pipelined_scheduling.py](https://github.com/iree-org/wave/pull/77/files#diff-1faec07204fed5bf51618ea52dcf5dd6916eb1bbdf967ec5265285e6cb2d31e6) it is `dict[Operation, GemmFourStageStage]`.
If you want to support different types of `operation_stage_table`s in this utils file you could use the following approach:
```python
from typing import TypeVar
from enum import Enum

ScheduleStage = TypeVar('ScheduleStage', bound=Enum)

def get_scheduling_stage(
    op: fx.Node, 
    operation_stage_table: dict[Operation, ScheduleStage]
) -> ScheduleStage:
```
We know the key is always the `Operation` Enum so we can directly use that. By using `TypeVar` this specifies that we expect a subclass of Enum as value in the dict and, crucially, the same subclass as return type of this function.

This also let's you drop your alias `ScheduleStage = Enum  # alias`

SourishW replied:
Thanks for your suggestion, I did this with the TypeVar as you suggested

martin-luecke commented on wave_lang/kernel/wave/scheduling/gemm_four_stage_pipelined_scheduling.py:
```diff
@@ -0,0 +1,112 @@
+                schedule[node] = next_stage.value
+                current_stage = next_stage
+            else:
+                # Node do not move contigously through stages.
```
```suggestion
                # Node does not move contiguously through stages.
```

martin-luecke commented on wave_lang/kernel/wave/scheduling/gemm_four_stage_pipelined_scheduling.py:
```diff
@@ -0,0 +1,112 @@
+
+    """
+
+    def mega_pipelined_scheduling(self, graph: fx.Graph, edges: list[Edge]):
+        """
+        Classify node to different stages. Based on it's stage,
+        program schedules clock for each node. This function also checks
+        that sorted node "contiguously" move between stages.
+        """
```
Why is this called "mega_pipelined"? What do you mean with "Based on it's stage, program schedules clock for each node." 
Could you please rework the name and comment to reflect better what is happening here.
This function is missing a return type annotation.

SourishW replied:
It was the old name I gave to scheduler type, I cleaned up the language to 

Classify node to different stages. Based on its stage,
        program schedules the node to a specific cycle for  each node.
        This function also checks that the sorted nodes move contiguously through
        expected stages.

martin-luecke commented on wave_lang/kernel/wave/scheduling/gemm_four_stage_pipelined_scheduling.py:
```diff
@@ -0,0 +1,112 @@
+    Operation.WRITE_GLOBAL: GemmFourStageStage.COMPUTE,
+}
+
+
+class GemmFourStageScheduler(BaseScheduler):
+    """
+    GEMM Four Stage Pipelined Scheduler
+
+    Convert vanilla schedule of:
+        for i = 0 to N:
+            a = READ_GLOBAL i
+            WRITE_SHARED a
+            barrier
+            b = READ_SHARED
+            COMPUTE b
```
Is this specific to GEMMs only? Does it make sense to simply call it `FourStageScheduler`?

raikonenfnu replied:
That's good, we actually have that check already implemented in schedule_reordering. But probably better to move it into this and also prefetch scheduler.

https://github.com/iree-org/wave/blob/main/wave_lang/kernel/wave/schedule_reordering.py#L164-L168

SourishW replied:
I will rename to FourStageScheduler and add the precondition

SourishW replied:
Done

martin-luecke commented on wave_lang/kernel/wave/scheduling/gemm_four_stage_pipelined_scheduling.py:
```diff
@@ -0,0 +1,112 @@
+    def next(self):
+
+        if self.value == 3:
+            return GemmFourStageStage(3)
```
```suggestion
            return self
```

SourishW replied:
Done

martin-luecke commented on wave_lang/kernel/wave/scheduling/gemm_four_stage_pipelined_scheduling.py:
```diff
@@ -0,0 +1,112 @@
+    # Helper function to get next stage from the current.
+    # If at stage 3 returns itself to prevent crash
+    # since it is final stage.
+    def next(self):
```
This needs type annotations. 
For Python3.11+ we could use `def next(self) → Self:`, otherwise  `def next(self) → "GemmFourStageStage":`

SourishW replied:
I used the latter

martin-luecke reviewed (changes requested):
Thank you for the PR!

Could you explain whether this is really only specific to Gemms or whether it might be usable for other kernels. In the case it is, I'd advocate for simply calling it `four_stage` everywhere.

Could you also explain which parts of the current multi-buffering implementation you think we could/should move to `loop_reconstruction`?
I think it might be worthwhile thinking about moving your `_heuristically_multi_buffer_shared_memory` to multi_buffering.py and simply calling it from loop_reconstruction (which already is a huge file and does more than "reconstructing" a loop). 


Otherwise, I left a few questions and suggestions.

SourishW commented on wave_lang/kernel/wave/scheduling/scheduler_utils.py:
```diff
@@ -0,0 +1,47 @@
+        edges: list[Edge],
+        resources: list[int],
+    ) -> None:
+        # assert False
```
Removed

SourishW commented on wave_lang/kernel/wave/scheduling/multi_buffering.py:
```diff
@@ -75,6 +121,7 @@ def _multi_buffer_memory_location(
     write_nodes: list[Write],
     reduction_axis: IndexSymbol,
     buffer_count: int,
+    scheduling_type: SchedulingType,
```
Just removed it thanks

SourishW commented on wave_lang/kernel/wave/scheduling/loop_reconstruction.py:
```diff
@@ -63,7 +68,28 @@ def update_node_index(
+    induction_variable: IndexSymbol,
+    reduction: Iterate,
+    node: fx.Node,
+):
```
Done

SourishW commented:
@raikonenfnu @harsh-nod @martin-luecke Thank you so much for your reviews, 
I have addressed the comments and made appropriate changes. 

@martin-luecke I was also able to make the multibuffering strategy scheduling type agnostic so that we don't need different logic for each and every single new scheduling strategy we introduce. 

This is ported your original code for multibuffering out, and this makes it so that we multibuffer only in one place in a convenient location for all scheduling type, not in 5 uniquely for each and every future scheduling type. It should work seamlessly for any new scheduling type introduced. 

Now instead of specifying `SchedulingType.MODULO_MULTI_BUFFERED`, you specify `SchedulingType.MODULO` and turn on the wave compile flag for `use_multi_buffering=True`, optionally specifying `multi_buffer_count` (default of 2)

@raikonenfnu let me know if you want me to move the buffer indexing based on fastest dimension to this PR or a future PR. 

I will work on a PR for indexing analysis now. If you want me to work on fastest dimension, I will work on that simultaneously and make a combined PR

martin-luecke reviewed (approved):
LGTM after implementing a check for the pre-conditions we have for this scheduling aproach

Thank you for the changes.

SourishW commented on wave_lang/kernel/wave/scheduling/scheduler_utils.py:
```diff
@@ -0,0 +1,49 @@
+        """
+        Returns the number of stages in the kernel of the pipelined loop.
+        """
+        max_cycle = max([t + 1 for t in self.schedule.values()])
```
Done (in 10 mins)

Hardcode84 commented on wave_lang/kernel/wave/scheduling/multi_buffering.py:
```diff
@@ -41,104 +88,33 @@ def multi_buffer(trace: CapturedTrace):
-    write_nodes: list[Write],
-    reduction_axis: IndexSymbol,
-    buffer_count: int,
+def _increase_dimensions_for_multibuffering(
```
`_increase_dimensions_for_multibuffering` currently multiplies each non-reduction dim by `buffer_count` which will cause issues if we have more than 1, I think we should multiply only one dim.

martin-luecke commented:
@Hardcode84 @SourishW I advocate for merging this PR with the current approach. We can investigate possibly switching to different multi-buffering buffer management afterwards. I remember looking at your suggestion of increasing the loop carried vars but had some problems I was not happy with. Possibly increased register pressure 🤔
The best hipblaslt kernels use this strategy so I went with it.

raikonenfnu commented:
> @Hardcode84 @SourishW I advocate for merging this PR with the current approach. We can investigate possibly switching to different multi-buffering buffer management afterwards. I remember looking at your suggestion of increasing the loop carried vars but had some problems I was not happy with. Possibly increased register pressure 🤔 The best hipblaslt kernels use this strategy so I went with it.

Agreed, but right now it still has failed schedule with the fully pipelined w/ global load to LDS

SourishW commented on wave_lang/kernel/wave/compile_options.py:
```diff
@@ -25,6 +25,8 @@ class WaveCompileOptions:
     schedule: bool = SchedulingType.NONE
     use_scheduling_barriers: bool = False
+    use_multi_buffering: bool = False
+    multi_buffer_count: int = 2  # if no buffer count specified do double buffer
```
Done

SourishW commented:
> > @Hardcode84 @SourishW I advocate for merging this PR with the current approach. We can investigate possibly switching to different multi-buffering buffer management afterwards. I remember looking at your suggestion of increasing the loop carried vars but had some problems I was not happy with. Possibly increased register pressure 🤔 The best hipblaslt kernels use this strategy so I went with it.
> 
> Agreed, but right now it still has failed schedule with the fully pipelined w/ global load to LDS

No more failed schedule with global load to LDS!

SourishW commented:
> Please fix the DCO

I just fixed it, sorry I missed this

Hardcode84 commented on wave_lang/kernel/wave/scheduling/four_stage_pipelined_scheduling.py:
```diff
@@ -0,0 +1,149 @@
+    COMPUTE = auto()
+    SCHEDULING_NOOP = -1
+
+    # Helper function to get next stage from the current.
```
this comment is no longer relevant

Hardcode84 commented on wave_lang/kernel/wave/scheduling/multi_buffering.py:
```diff
@@ -41,114 +107,45 @@ def multi_buffer(trace: CapturedTrace):
     # Find reads and writes operating on shared memory
-    reads = []
-    writes = []
+    reads_and_writes = list()
```
`reads_and_writes = []`

Hardcode84 commented on wave_lang/kernel/wave/compile_options.py:
```diff
@@ -25,6 +25,9 @@ class WaveCompileOptions:
     # === Scheduling options ===
     schedule: bool = SchedulingType.NONE
     use_scheduling_barriers: bool = False
+    multi_buffer_count: Optional[int] = (
```
weird formatting, move comment to separate line

SourishW replied:
Just added to the kernel hash

rohan-tan-bhowmik commented on wave_lang/kernel/wave/scheduling/multi_buffering.py:
```diff
@@ -23,10 +21,78 @@
+        new_node_idx = get_dict_with_updated_key(
+            new_node_idx, dim, dim * multi_buffer_count
+        )
+        if is_read_write:
+            new_node.index = new_node_idx
+        else:
+            new_node.dst_index = new_node_idx
+            return
```
Tiny tiny suggestion for clarity and brevity

```suggestion
        if not is_read_write:
            new_node.dst_index = new_node_idx
            return
            
        new_node.index = new_node_idx
```

paulzzy replied:
LGTM
